### ISSUE TITLE
Large X shaped pattern tiled across images

### ISSUE THREAD

**Issue by mburge** [aaronwmorris/indi-allsky#1187]
Any idea where this large X shaped pattern tiled across the images is originating?  I am stacking 5 images (taking the maximum) with registration and my guess is it is originating in that process.  The raw images are being captured in 16 bit from an ASI294 with 60 second exposures. I'll try to chase it down tomorrow but thought you might have an idea off hand. I'm saving images as jpg's using the default settings. I attached a sample in PNG format.

![whyXs_ccd1_20240311_215549](https://github.com/aaronwmorris/indi-allsky/assets/3086001/c1838558-3fbc-45b9-b95b-422f67380602)

**Comment by aaronwmorris** [comment]
Are you using the CLAHE for contrast enhancement?

**Comment by mburge** [comment]
No, it is not enabled, in addition to the stack with registration, the only things on are AWB, saturation at 1.5, average neutral SCNR, and stretching. All at defaults. I defocused the camera to check if it was in the light path, but no luck, here's an externally contrast enhanced detail:

![image](https://github.com/aaronwmorris/indi-allsky/assets/3086001/d18c8d1a-8d1e-42bb-9e4a-bcfef1cbc982)

**Comment by mburge** [comment]
Looks like it is an artifact of the registration process. I disabled all stacking. No artifacts. Enabled stacking 5 with maximum but w/out registration. Got images with star trails consisting of 5 stars, as you'd expect with maximum and no registration. Enabled registration and artifacts returned. My guess is that when solving for the transformation matrix it is iteratively alternating between close solutions, maybe just because of precision. Kind of neat actually, that is why the structure is so regular. I'll look tomorrow. By the way, love your system.

**Comment by aaronwmorris** [comment]
Thank you!

I could barely see the pattern on my laptop last night (but I could see it), but I can see it much better on my workstation.

I have used registration quite a few times, although, I rarely leave it on.  It has to be a clear night.  However, I have never seen that.  My cameras are not as high resolution as the ASI294.

https://github.com/quatrope/astroalign is the module used for registration.  I will be the first to admit that the registration can be less than perfect.  All sky images have enough distortion and enough variations due to planes and satellites that misalignments are very common.  After I implemented the filter to check for rotations exceeding a standard deviation, misalignments are less common now.

I wonder if this pattern is some interaction due to the quad bayer matrix on the IMX294.

**Comment by mburge** [comment]
I had a chance to run some tests and it is definitely a result of registration.

Split screen with registration
![image](https://github.com/aaronwmorris/indi-allsky/assets/3086001/5b90ea56-8c79-446a-be6e-fd04aa1a882a)

Split screen with just stacking
![image](https://github.com/aaronwmorris/indi-allsky/assets/3086001/56ec6de5-dbc1-4490-8f6b-4e5402336ce5)

I haven't found where it is yet, I think you're right, it's an artifact of the quad mask / 16 bit format. Solution for now is to just turn off stacking and registration.

![image](https://github.com/aaronwmorris/indi-allsky/assets/3086001/e6e3df1f-caa2-495c-9726-4eb29838fc7a)

**Comment by jschmied** [comment]
A example with imx477. 
![ccd1_20241126_185941](https://github.com/user-attachments/assets/f766002e-0208-4969-a317-4ebdabe42a14)

**Comment by aaronwmorris** [comment]
I still have not been able to reproduce this.  Are you using CLAHE?

**Comment by jschmied** [comment]
I don't use any enhancements. Only SCNR. Pattern changes with number of registered pictures. Maybe I should ask in the astroalign project ..

**Comment by aaronwmorris** [comment]
@jschmied Did you happen to have your camera sitting in front of a screen door or something like that?

**Comment by jschmied** [comment]
No, pictures without stacking are without pattern of course. I suspect its a product of imperfect sensor, debayer and stacking. How can I keep the dng files saved by libcamera? Export RAW is exporting after debayer. I would like to check if processing with siril or deepskystacker produces pattern also. Maybe I find a option causing this. Pattern looks like Moire of superimposed gitter:

![Bildschirmfoto 2024-11-28 um 08 28 29](https://github.com/user-attachments/assets/5bb9c46c-83c8-41b7-ba44-1af29036526c)

**Comment by aaronwmorris** [comment]
The FITS processor is probably the best way to test image changes faster than others.  Changes to the code are immediately available without having to restart anything.

The FITS processor uses the same exact code methods as the image processing during capture.  All of the processing magic happens in `indi_allsky/processing.py` (and included modules).

I have had this thought previously, however, I have not investigated yet.  You can see in the order of operations https://github.com/aaronwmorris/indi-allsky/wiki/Image-order-of-operations that stacking occurs before debayering and that could be the cause of the moire

**Comment by jschmied** [comment]
Simple stacking needs to be moved after debayer. Otherweise offset and rotation aligns different colors into same pixel.

Just swaping the function calls is not working because stack() is populating self.image and debayer() gets AttributeError in flask/views.py", line 5777 -> processing.py", line 1061

Also other participants in stacking (self.image_list) must be debayered.

Have to get deeper into the code ...

**Comment by aaronwmorris** [comment]
Merged #1699 which performs the debayering before stacking.  The astroalign module does indeed support registering color images.

@jschmied @mburge Please let me know if this corrects the issue with the moire pattern.

**Comment by jschmied** [comment]
Thanks for your effort! I installed it but its very cloudy the next days.

**Comment by jschmied** [comment]
It's fixed! Thanks for your effort.

Stacked (I need to do darks, you can see trails of hot pixels)
![Bildschirmfoto 2024-12-06 um 08 13 33](https://github.com/user-attachments/assets/a4d9773f-66ff-44ec-80ea-cb0766cb3720)

Original (more noise)
![Bildschirmfoto 2024-12-06 um 08 15 49](https://github.com/user-attachments/assets/8c4f98fa-c9ad-4390-8c5c-2dbed10e8656)

Stacked: (Detla, Gamma, Eta Cassiopeiae)
![Bildschirmfoto 2024-12-06 um 08 20 28](https://github.com/user-attachments/assets/88c5205d-d3c4-4e3c-9a47-61c3025a7169)

Unstacked:
![Bildschirmfoto 2024-12-06 um 08 21 34](https://github.com/user-attachments/assets/bcd25ad6-8fbd-4d1e-a588-afe06662e2d4)

**Comment by aaronwmorris** [comment]
You may already be using this, but I recommend using Maximum instead of Average for the stacking of color sensors.  

This is kind of unique to all sky cameras, but when you have achieved critical focus, most stars will only illuminate a single pixel on the camera.  Due to the bayer filter, the star will sometimes just be red, green, or blue.  When a star moves between frames, it might be red one frame and green in another.  The averaging occurs in each color channel, so if you had a star run across 3 different frames, the overall luminance can be reduced by up to 66% (for the star).

For example, the average of

- [255, 0, 0]
- [0, 255, 0]
- [0, 0, 255]

is

- [85, 85, 85]

Of course, this is all a matter of taste.  :-)

**Comment by jschmied** [comment]
Sounds reasonable. I already added some code because it tries to stack pictures without stars and fails after 7 seconds to find a match. I'll send a PR next week.
